pkg/openstack-cinder/config/configsync.go

```diff
@@ -18,56 +16,107 @@ import (
-	kubeClient kubernetes.Interface,
-	informers v1helpers.KubeInformersForNamespaces,
-	configInformers configinformers.SharedInformerFactory,
-	guestNamespace string,
-	resyncInterval time.Duration,
-	eventRecorder events.Recorder) factory.Controller {
-
-	// Read configmap from user-managed namespace and save the translated one
-	// to the operator namespace
-	configMapInformer := informers.InformersFor(util.OpenShiftConfigNamespace)
-	c := &ConfigSyncController{
-		operatorClient:       operatorClient,
-		kubeClient:           kubeClient,
-		configMapLister:      configMapInformer.Core().V1().ConfigMaps().Lister(),
-		infrastructureLister: configInformers.Config().V1().Infrastructures().Lister(),
-		guestNamespace:       guestNamespace,
-		eventRecorder:        eventRecorder.WithComponentSuffix("ConfigSync"),
-	}
-	return factory.New().WithSync(c.sync).ResyncEvery(resyncInterval).WithSyncDegradedOnError(operatorClient).WithInformers(
-		operatorClient.Informer(),
-		configMapInformer.Core().V1().ConfigMaps().Informer(),
-	).ToController("ConfigSync", eventRecorder)
+// NewConfigSyncController creates a new instance of ConfigSyncController
+func NewConfigSyncController(c *clients.Clients, isHypershift bool) (factory.Controller, error) {
+	var configMapInformer coreinformers.ConfigMapInformer
+	var configSources []configSource
+	var configDestinations []configDestination
+
+	if !isHypershift {
+		// If we're in a standalone deployment, we watch the openshift-config namespace...
+		configMapInformer = c.GetConfigMapInformer(util.OpenShiftConfigNamespace)
+
+		infra, err := c.GetInfraInformer().Lister().Get(infrastructureResourceName)
+		if err != nil {
+			return nil, err
+		}
+
+		// ...and try to source potential user-provided configuration from one of two config maps
+		configSources = []configSource{
+			// First, we try to retrieve from the Cinder CSI-specific config map
+			{util.OpenShiftConfigNamespace, "cinder-csi-config", "config"},
+			// Failing that, we attempt to retrieve from the cloud provider-specific config map
+			// TODO(stephenfin): We should stop retrieving this once Installer creates the new
+			// config, which will allow us to simplify things somewhat here
+			{util.OpenShiftConfigNamespace, infra.Spec.CloudConfig.Name, infra.Spec.CloudConfig.Key},
+		}
+
+		// ...while saving to the only cluster we have
+		configDestinations = []configDestination{
+			{c.ControlPlaneNamespace, util.CinderConfigName, c.KubeClient},
+		}
+	} else {
+		// If we're in a hypershift deployment, we watch the clusters-${NAME} namespace
+		// (on the control plane cluster)...
+		configMapInformer = c.GetControlPlaneConfigMapInformer(c.ControlPlaneNamespace)
+
+		// ...and we don't currently support user-defined configuration so there's nothing to set
+		// for that
+
+		// ...but we now have two clusters to save to
+		configDestinations = []configDestination{
+			{c.ControlPlaneNamespace, util.CinderConfigName, c.ControlPlaneKubeClient},
```

Comment:
is it really needed in the control plane namespace?
I took a look on the assets generated and the DaemonSet that runs on the guest cluster seems to be the only one that needs it https://github.com/openshift/csi-operator/pull/302/files#diff-dece87039665ecf71f9003b48e52ce76685b59171f7c8e74ae0110f72b0526acR197.

Reply:
The diff for `controller.yaml` is hidden by default so you need to expand that. When you do, you'll see it's used in the controller also.

https://github.com/openshift/csi-operator/pull/302/files#diff-8114e4393b7e73a8d7b1a309b1d0688968e54ffe7b996087cd410b780d386731R412